### PR TITLE
feat: `aio app pack` command

### PR DESCRIPTION
**Pull request by shazron**
(this is based on the WIP PoC for the aio app package command)

- depends on https://github.com/adobe/aio-cli-plugin-api-mesh/pull/61 (worked around, for now)
    -  the PR has been merged, but not released in a new plugin release

## How Has This Been Tested?

- npm test

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] I have signed the [Adobe Open Source CLA](http://opensource.adobe.com/cla.html).
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by shazron** [comment]
I'm on Windows, and it was all passing -- had to reclone the repo and, npm i and npm test again to see the Windows errors (related to line endings).

File: src/commands/app/pack.js [hunk 1]
```diff
@@ -66,15 +64,14 @@ class Pack extends BaseCommand {
...
+    // doing this before zip so other things can be added to the zip
+    await this.config.runHook('post-pack', { appConfig, artifactsFolder: DEFAULTS.ARTIFACTS_FOLDER })
 
     // 4. zip package phase
     this.log(`Zipping package artifacts folder '${DEFAULTS.ARTIFACTS_FOLDER}' to '${flags.output}'...`)
     await fs.remove(flags.output)
     await this.zipHelper(DEFAULTS.ARTIFACTS_FOLDER, flags.output)
 
-    // ACNA-2038
-    // TODO: fire post-pack event for Events
```
Comment (sangeetha5491) [inline src/commands/app/pack.js]: we wont need this

Comment (shazron) [reply src/commands/app/pack.js]: based on the internal Slack thread discussion, I'll resolve this.

**Review by moritzraho — APPROVED** [review]
lgtm, great mocks

File: src/commands/app/pack.js [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+    for (const configFile of configFiles) {
+      const configFilePath = path.join(DEFAULTS.ARTIFACTS_FOLDER, configFile.file)
+      const { values } = loadConfigFile(configFilePath)
+
+      const runtimeManifest = getObjectValue(values, configFile.key)
+      for (const [, pkgManifest] of Object.entries(runtimeManifest.packages)) {
+        // key is the package name (unused), value is the package manifest. we iterate through each package's "actions"
+        for (const [, actionManifest] of Object.entries(pkgManifest.actions)) {
+          // key is the action name (unused), value is the action manifest. we add the "code-download: false" annotation
+          actionManifest.annotations['code-download'] = false
```
Comment (moritzraho) [inline src/commands/app/pack.js:236]: nitpicking but should we keep true if the user explicitly sets it to true? (although I wouldn't see why)

Comment (shazron) [reply src/commands/app/pack.js:236]: `aio app pack` as proposed is meant for ISV use, which means `code-download` annotation will always need to be false, which is a requirement. If we want to honor the existing annotation, that is possible, but not sure if that was the intended design. Will bring it up in Slack.

Comment (shazron) [reply src/commands/app/pack.js:236]: Resolved based on internal Slack discussion: we always override with `false`. If a developer wants to make their ISV app-package downloadable on Exchange, then they can choose to do so in DD during the app submission.

**Review by purplecabbage — APPROVED** [review]
Looks good, I left a couple nits.
We may need to move around the pre-pack/post-pack hook timing.
Also, should we skip cleaning up the output dir? While we are developing/integrating it might be good to have this easily inspectable.  We do a similar thing with zipped actions where we do not discard the files so developers can easily see what is going in the zip.

File: src/commands/app/pack.js [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+      $id: 'https://adobe.io/schemas/app-builder-templates/1',
+      application,
+      extensions,
+      workspaces,
+      apis,
+      meshConfig,
+      runtime: true // always true for App Builder apps
+    }
+
+    await writeFile(
```
Comment (purplecabbage) [inline src/commands/app/pack.js:135]: is it yaml or json? I know yaml includes json, but this looks a little confusing.

Comment (shazron) [reply src/commands/app/pack.js:135]: we write out yaml, as that is what `yaml.dump` outputs, and takes json as input.

File: src/lib/app-helper.js [hunk 4]
```diff
@@ -505,7 +505,32 @@ const createWebExportFilter = (filterValue) => {
...
 
+/**
+ * Get property from object with case insensitivity.
+ *
+ * @param {object} obj the object to wrap
+ * @param {string} key the key
+ * @private
+ */
+function getObjectProp (obj, key) {
+  return obj[Object.keys(obj).find(k => k.toLowerCase() === key.toLowerCase())]
```
[the remark this answers is not in the document]
Comment (shazron) [reply src/lib/app-helper.js:516]: there is actually two lowerCase() calls here, and there is point in in having the local var I feel

File: src/commands/app/pack.js [hunk 5]
```diff
@@ -0,0 +1,266 @@
...
+    // change the cwd if necessary
+    if (args.path !== '.') {
+      const resolvedPath = path.resolve(args.path)
+      process.chdir(resolvedPath)
+      aioLogger.debug(`changed current working directory to: ${resolvedPath}`)
+    }
+
+    // 1. create artifacts phase
+    this.log('Creating package artifacts...')
+    await fs.remove(DEFAULTS.ARTIFACTS_FOLDER)
```
[the remark this answers is not in the document]
Comment (shazron) [reply src/commands/app/pack.js]: fixed in https://github.com/adobe/aio-cli-plugin-app/pull/650/commits/7324dd4dd2a6564f7e10f3a80d4d3c58352a4415

**Comment by shazron** [comment]
> Looks good, I left a couple nits. We may need to move around the pre-pack/post-pack hook timing. 

@purplecabbage Is this a change we need in this PR? What changes are needed?

>Also, should we skip cleaning up the output dir? While we are developing/integrating it might be good to have this easily inspectable. We do a similar thing with zipped actions where we do not discard the files so developers can easily see what is going in the zip.

We don't clean it up after the command is done - we only clean up at the start of the command

**Comment by shazron** [comment]
I'll merge this first thing tomorrow and do more testing, which will include #662

**Pull request merged**